tools/wake-format/emitter.cpp

```diff
@@ -403,7 +426,72 @@ wcl::optional<wcl::doc> Emitter::walk_data(ctx_t ctx, CSTElement node) {
 }
 
 wcl::optional<wcl::doc> Emitter::walk_def(ctx_t ctx, CSTElement node) {
-  return walk_placeholder(ctx, node);
+  ASSERT_TOKEN(node, CST_DEF);
+  wcl::doc_builder builder;
+  CSTElement child = node.firstChildElement();
+
+  // def shouldn't follow anything other than indentation.
+  if (ctx.width > ctx.nest_level * space_per_indent) {
+    builder.append(newline(ctx));
+  }
+
+  // def
+  ASSERT_TOKEN(child, TOKEN_KW_DEF);
+  builder.append("def");
+  child.nextSiblingElement();
+
+  // ws
+  ASSERT_TOKEN(child, TOKEN_WS);
+  builder.append(space());
+  child.nextSiblingElement();
+
+  // CST_ID, CST_APPLY, TODO: others?
+  assert(!child.empty());
+  assert(child.id() == CST_ID || child.id() == CST_APP);
+  {
+    auto new_child = walk_node(ctx.sub(builder), child);
+    assert(new_child);
+    builder.append(*new_child);
+  }
+  child.nextSiblingElement();
+
+  // ws
+  ASSERT_TOKEN(child, TOKEN_WS);
+  builder.append(space());
+  child.nextSiblingElement();
+
+  //  =
+  ASSERT_TOKEN(child, TOKEN_P_EQUALS);
+  builder.append("=");
+  child.nextSiblingElement();
+
+  // optional ws/nl
+  CONSUME_WS_NL(child);
+
+  // rhs
+  {
+    auto new_child = walk_node(ctx.sub(builder), child);
+    assert(new_child);
+    wcl::doc new_doc = *new_child;
+    new_doc = space().concat(new_doc);
+    if (builder.last_width() + new_doc.first_width() + ctx.width <= max_column_width) {
```

Comment:
If the builder has a newline this isn't the right calculation, we probably need a helper function to get this correct....but all of the info is *here* finally which is fucking amazing.

Reply:
how so? if builder has a newline then `last_width()` is 0

Reply:
It might have a newline in the *middle*....ok to be fair we know there's no newline in this case but I'm trying to think more generically about it. We want a function that computes "will adding this doc overflow this builder's last line?" and this gets that wrong for some builders since if the builder has a newline you don't want to include `ctx.width`

```
bool fits(const wcl::doc_builder& bdr, ctx_t ctx, wcl::doc new_doc) {
  if (bdr.has_newline()) {
    return bdr.last_width() + new_doc.first_width() <= max_column_width;
  } else {
    return bdr.last_width() + new_doc.first_width() + ctx.width <= max_column_width
  }
}
```

Reply:
ah right, thanks!